File: agentaid-claude-service/services/geocodingService.js
import dotenv from 'dotenv';

dotenv.config();

// You can use OpenCage (free) or Google Maps
const GEOCODING_PROVIDER = process.env.GEOCODING_PROVIDER || 'opencage'; // or 'google'
const OPENCAGE_API_KEY = process.env.OPENCAGE_API_KEY;
const GOOGLE_MAPS_API_KEY = process.env.GOOGLE_MAPS_API_KEY;

// OpenCage Geocoding (Free tier: 2,500 requests/day)
async function geocodeWithOpenCage(address) {
  try {
    const encodedAddress = encodeURIComponent(address);
    const url = `https://api.opencagedata.com/geocode/v1/json?q=${encodedAddress}&key=${OPENCAGE_API_KEY}&limit=1`;
    
    const response = await fetch(url);
    const data = await response.json();
    
    if (data.results && data.results.length > 0) {
      const result = data.results[0];
      return {
        latitude: result.geometry.lat,
        longitude: result.geometry.lng,
        formatted_address: result.formatted,
        confidence: result.confidence,
        components: {
          city: result.components.city || result.components.town,
          state: result.components.state,
          country: result.components.country,
          postcode: result.components.postcode
        }
      };
    }
    
    return null;
  } catch (error) {
    console.error('OpenCage geocoding error:', error);
    return null;
  }
}

// Google Maps Geocoding (Paid, but more accurate)
async function geocodeWithGoogle(address) {
  try {
    const encodedAddress = encodeURIComponent(address);
    const url = `https://maps.googleapis.com/maps/api/geocode/json?address=${encodedAddress}&key=${GOOGLE_MAPS_API_KEY}`;
    
    const response = await fetch(url);
    const data = await response.json();
    
    if (data.results && data.results.length > 0) {
      const result = data.results[0];
      return {
        latitude: result.geometry.location.lat,
        longitude: result.geometry.location.lng,
        formatted_address: result.formatted_address,
        place_id: result.place_id,
        components: {
          city: result.address_components.find(c => c.types.includes('locality'))?.long_name,
          state: result.address_components.find(c => c.types.includes('administrative_area_level_1'))?.long_name,
          country: result.address_components.find(c => c.types.includes('country'))?.long_name,
          postcode: result.address_components.find(c => c.types.includes('postal_code'))?.long_name
        }
      };
    }
    
    return null;
  } catch (error) {
    console.error('Google geocoding error:', error);
    return null;
  }
}

// Fallback: Simple geocoding using Nominatim (OpenStreetMap - Free, no API key)
async function geocodeWithNominatim(address) {
  try {
    const encodedAddress = encodeURIComponent(address);
    const url = `https://nominatim.openstreetmap.org/search?q=${encodedAddress}&format=json&limit=1`;
    
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'AgentAid-DisasterResponse/1.0' // Required by Nominatim
      }
    });
    
    const data = await response.json();
    
    if (data && data.length > 0) {
      const result = data[0];
      return {
        latitude: parseFloat(result.lat),
        longitude: parseFloat(result.lon),
        formatted_address: result.display_name,
        confidence: 0.8, // Nominatim doesn't provide confidence
        components: {
          city: result.address?.city || result.address?.town,
          state: result.address?.state,
          country: result.address?.country,
          postcode: result.address?.postcode
        }
      };
    }
    
    return null;
  } catch (error) {
    console.error('Nominatim geocoding error:', error);
    return null;
  }
}

// Main geocoding function
export async function geocodeAddress(address) {
  if (!address || address.trim().length < 5) {
    console.log('⚠️  Address too short for geocoding');
    return null;
  }
  
  console.log(`🌍 Geocoding address: "${address}"`);
  
  try {
    let result = null;
    
    // Try primary provider
    if (GEOCODING_PROVIDER === 'google' && GOOGLE_MAPS_API_KEY) {
      result = await geocodeWithGoogle(address);
    } else if (GEOCODING_PROVIDER === 'opencage' && OPENCAGE_API_KEY) {
      result = await geocodeWithOpenCage(address);
    } else {
      // Fallback to free Nominatim
      console.log('📍 Using Nominatim (free, no API key)');
      result = await geocodeWithNominatim(address);
    }
    
    if (result) {
      console.log(`✅ Geocoded: ${result.latitude}, ${result.longitude}`);
      return result;
    }
    
    console.log('⚠️  Could not geocode address');
    return null;
    
  } catch (error) {
    console.error('❌ Geocoding error:', error);
    return null;
  }
}

// Calculate distance between two coordinates (in kilometers)
export function calculateDistance(lat1, lon1, lat2, lon2) {
  const R = 6371; // Earth's radius in km
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  
  const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) *
    Math.sin(dLon / 2) * Math.sin(dLon / 2);
  
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  const distance = R * c;
  
  return distance;
}

function toRad(degrees) {
  return degrees * (Math.PI / 180);
}

// Find nearest location from a list
export function findNearestLocation(targetLat, targetLon, locations) {
  if (!locations || locations.length === 0) return null;
  
  let nearest = null;
  let minDistance = Infinity;
  
  for (const location of locations) {
    if (location.latitude && location.longitude) {
      const distance = calculateDistance(
        targetLat,
        targetLon,
        location.latitude,
        location.longitude
      );
      
      if (distance < minDistance) {
        minDistance = distance;
        nearest = { ...location, distance };
      }
    }
  }
  
  return nearest;
}

// Batch geocode multiple addresses
export async function batchGeocode(addresses) {
  const results = [];
  
  for (const address of addresses) {
    const result = await geocodeAddress(address);
    results.push({
      address: address,
      geocoded: result
    });
    
    // Rate limiting: wait 1 second between requests (for free APIs)
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  
  return results;
}

File: services/geocoding_service.py
"""
Geocoding Service for AgentAid
Handles address geocoding using OpenStreetMap Nominatim
"""

import requests
import time
from typing import Dict, Any, Optional, List
import math

class GeocodingService:
    def __init__(self):
        self.base_url = "https://nominatim.openstreetmap.org/search"
        self.headers = {
            'User-Agent': 'AgentAid-DisasterResponse/1.0'  # Required by Nominatim
        }
        self.rate_limit_delay = 1.0  # Seconds between requests

    def geocode_address(self, address: str) -> Optional[Dict[str, Any]]:
        """Geocode an address to coordinates"""
        if not address or len(address.strip()) < 5:
            print("⚠️ Address too short for geocoding")
            return None

        print(f"🌍 Geocoding address: \"{address}\"")

        try:
            params = {
                'q': address,
                'format': 'json',
                'limit': 1,
                'addressdetails': 1
            }

            response = requests.get(
                self.base_url,
                params=params,
                headers=self.headers,
                timeout=10
            )

            if response.status_code == 200:
                data = response.json()

                if data and len(data) > 0:
                    result = data[0]

                    geocoded = {
                        'latitude': float(result['lat']),
                        'longitude': float(result['lon']),
                        'formatted_address': result['display_name'],
                        'confidence': 0.8,  # Nominatim doesn't provide confidence
                        'components': {
                            'city': result.get('address', {}).get('city') or result.get('address', {}).get('town'),
                            'state': result.get('address', {}).get('state'),
                            'country': result.get('address', {}).get('country'),
                            'postcode': result.get('address', {}).get('postcode')
                        }
                    }

                    print(f"✅ Geocoded: {geocoded['latitude']}, {geocoded['longitude']}")
                    return geocoded

            print("⚠️ Could not geocode address")
            return None

        except Exception as e:
            print(f"❌ Geocoding error: {e}")
            return None

    def batch_geocode(self, addresses: List[str]) -> List[Dict[str, Any]]:
        """Geocode multiple addresses with rate limiting"""
        results = []

        for address in addresses:
            result = self.geocode_address(address)
            results.append({
                'address': address,
                'geocoded': result
            })

            # Rate limiting: wait between requests
            time.sleep(self.rate_limit_delay)

        return results

    def calculate_distance(self, lat1: float, lon1: float, lat2: float, lon2: float) -> float:
        """Calculate distance between two coordinates in kilometers"""
        R = 6371  # Earth's radius in km

        dlat = math.radians(lat2 - lat1)
        dlon = math.radians(lon2 - lon1)

        a = (math.sin(dlat / 2) * math.sin(dlat / 2) +
             math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
             math.sin(dlon / 2) * math.sin(dlon / 2))

        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
        distance = R * c

        return distance

    def find_nearest_location(self, target_lat: float, target_lon: float, locations: List[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
        """Find nearest location from a list of locations"""
        if not locations:
            return None

        nearest = None
        min_distance = float('inf')

        for location in locations:
            if 'latitude' in location and 'longitude' in location:
                distance = self.calculate_distance(
                    target_lat,
                    target_lon,
                    location['latitude'],
                    location['longitude']
                )

                if distance < min_distance:
                    min_distance = distance
                    nearest = {**location, 'distance_km': distance}

        return nearest

    def check_health(self) -> bool:
        """Check if geocoding service is healthy"""
        try:
            # Test with a simple address
            test_result = self.geocode_address("San Francisco, CA")
            return test_result is not None
        except Exception:
            return False
